### PR TITLE
Loading/Error handling for Authenticated landing page

### LINKED ISSUE #121: [App] Loading State - Full Content
## Expected Behavior
Create a CFPB/SBL loading state that fills the current content section (e.g. User Profile, View Institutions, etc)

## Accepted Criteria
- Deployed on AWS
- Approval by a designer (i.e. Natalia, Dan)

### LINKED ISSUE #157: [Loading/Error Handling] Authenticated landing page
- [ ] Loading
  - [ ] Full content loading screen
- [ ] Errors - Institution data
  - [ ] No name? -> Render LEI
  - [ ] No LEI? -> Notify user to contact SBL Help
  - [ ] No associated institutions? -> Would be caught by ProtectedRoute checks

### LINKED ISSUE #158: [Loading/Error] View institution details
- [x] Loading
  - [x] fetchAssociatedInstitutions -> Full content loading screen
- [x] Errors - Institution data
  - [x] fetchAssociatedInstitutions fails? -> Global 500 page w/ error message

### LINKED ISSUE #159: [Loading/Error Handling] View User Profile
- [ ] Loading
  - [ ] fetchUserProfile -> Full content loading screen
- [ ] Errors - Institution data
  - [ ] fetchUserProfile fails? -> Global 500 page w/ error message

### PR DESCRIPTION
**Pull request by meissadia**
Closes #121 
Closes #157 
Closes #158 
Closes #159 

## Changes

- Enables "full content loading"
  - Provide /loading route for easy examination of Loading state
  - Addresses flexbox page-width issues on Privacy Act, etc
  - Creates `LoadingContent` (to be used by subpages) and `LoadingPage` (for use with React.Suspense) components
  - Refactors generation of Header authentication links into a custom hook (`useHeaderAuthLinks`) to avoid flashing during loading
  - Does not display Header Auth links until Auth is fully loaded to avoid flickering
  - Displays `Loading...` text and the associated icon in an H1
- Implements Loading for the `Authenticated Landing` page
  - Refactors data fetching in `AuthenticatedLanding` to live at the top level, so that we can show the whole page as loading while we wait for details to download
  - Replaces all loading instances of `<LoadingOrError>` with `<LoadingContent />` 
- Implements Loading and Error handling for the `View User Profile` and `View Institution Profile` pages
  - Replaces all loading instances of `<LoadingOrError>` with `<LoadingContent />` 
  - Redirects to /500 for data fetching errors
- Updates error handling, where it already exists, to use the /500 route
  - Provides a `useError500` hook to encapsulate error routing logic
  - Replaces all error instances of `<LoadingOrError>` with `useError500` 
  - Route users to `/500` when a server/fetch error occurs
- Display field
  - Render `Not available` for any field that does not have a `value` to display

## How to test this PR

1. Loading: Open http://localhost:8899/loading
2. Error - Unable to fetch: Set `ReviewInstitutions.tsx` line 23 to `true`
3. Warning - No associated institutions: Set `ReviewInstitutions.tsx` line 32 to `false`

#### All Good
AllGood

#### Warning - No associated institutions
warning-field-level

#### Error - Unable to fetch
error-field-level

#### `Not available` for *any* field without a value
Not-available

### Full content loading
Screenshot 2024-01-19 at 3 04 31 PM

## TODO
- [x] Verify with team that top-level fetching is the approach we want to follow
- [x] Figure out what to do about "unknown" type for useQuery's `TError`
  - Via test I've determined that `TError.message` will provide some text content but ultimately, based on the presence of the error and the context, we'll have to provide our own messaging.
- [x] Verify with Design team that the Alert's used to show "No institutions" and "Fetch error" look good

### CODE REVIEW

**Comment by meissadia** [comment]
Question about "No associated institutions" scenario:
- Is this even technically possible?
- Could a user remove all of their associated institutions?
- Could our Support team accidentally remove all of a user's associated institutions?

**Comment by natalia-fitzgerald** [comment]
@dan-padgett @angelcardoz 
If @meissadia confirms that these warnings or errors are possible we will have to decide how to handle the language in the text block above this field-level alert, the language in the alert, and also identify and present the next step a user should take. 

The current language is contradictory: 
- Your user profile has been successfully associated with the following financial institutions...
- You have no associated institutions

**Comment by billhimmelsbach** [comment]
I really appreciate this discussion @natalia-fitzgerald and @meissadia! I think we might need to clarify a bit about what problems are possible and the issues with trying to guess what errors we should convey to the user.

Screenshot 2024-01-19 at 4 49 16 PM

When @meissadia, @shindigira, and I were first brainstorming the different error states that were possible due to API calls failing in various ways, it was great to think about "if an institution came back from `getAssociatedInstitutions` with an LEI but not a name, could we still render this link to their View Financial Institution Details page on their Authenticated Landing page?" I think that error brainstorming was very helpful, but as this PR discussion is surfacing, from a user and design perspective it might be good to take a step back.

If an institution comes in from a successful `GET` from the `getAssociatedInstitutions` endpoint and the institution has a LEI but not a name, could we render that link to their View Financial Institution Details page? Totally. But since someone with an LEI in GLIEF must register with a name also, then not having a name come back from that endpoint speaks to a server error or an error in how we're interacting with the backend. To a user who has an LEI but no name associated with their institution, they should probably just see a "Error loading institution details; please contact SBL Help" with a link to SBL Help's salesforce form.

Similarly and speaking to @meissadia's point about "Is this even technically possible.", if a user gets to the authenticated landing page and doesn't have any associated institutions coming back from `getAssociatedInstitutions` that means that somehow their institutions array on `Get Me` is populated which is a requirement to get to this page, but that data is not being returned in `getAssociatedInstitutions` . That's another server error or error with how we're communicating with the backend so probably the user should just see: "Error loading institution details; please contact SBL Help".

I chatted a bit with @natalia-fitzgerald about this, but this speaks to a general pattern I'd like to adopt for these components that depend on API calls to render. Let's generalize all errors coming from this component into one message, preferably in a red alert error box that draws the user's attention with: "Error loading institution details; please contact SBL Help" with a link to SBL Help.

One way we could do this technically for all use cases, would be to put the dependent API call `getAssociatedInstitutions` into the component itself, and then wrap the component in a react Error Boundary that renders the red alert error box if it catches an error. Then if any errors occur inside the component, they get caught and people see the generalized error message. If later in testing we find that we want to be more specific with error text, then after MVP we could adjust the Error Boundary to render different messages as necessary.

Seems like a pretty good pattern that's repeatable across components? Let me know what you think y'all, and thanks for the discussion! 👍

**Comment by billhimmelsbach** [comment]
TL;DR:
- I think some non-critical components that depend on API calls to meaningfully render should be wrapped in Error Boundaries. These Error Boundaries should display fairly generic error text in red alert error boxes that direct users to SBL Help.
- Errors that happen on the page level or unexpected errors go to 500 pages.
- For later when we're working in the filing app: individual form pages in the filing app should use `zod` to verify form content before continuing onto the next page on the frontend. The backend should also send back errors if the form content contains errors as we submit each page of the draft filing to the backend to be saved.

**Comment by natalia-fitzgerald** [comment]
@billhimmelsbach 
Thanks for your notes. I went ahead and mocked this up for the authenticated landing page and the View your user profile page. I tried out two positioning locations - top of the page versus specific page location where the data didn't load. 

### Top of the content area
| Authenticated landing page | View your user profile |
|------------|------------|
Group 378|MVP_ View your user profile (error loading financial institution details) - v1|

### Specific page location
| Authenticated landing page | View your user profile |
|------------|------------|
|MVP_ Shared platform (authenticated) - error state v2|MVP_ View your user profile (error loading financial institution details) - v2|

If we want to provide further details we can include body copy in the alert. 
Screenshot 2024-01-19 at 10 20 07 PM

**Comment by billhimmelsbach** [comment]
@meissadia Is this one good to go?

**Comment by billhimmelsbach** [comment]
@meissadia @natalia-fitzgerald The latest is currently on the dev environment to aid review. 👍

**Comment by meissadia** [comment]
Discussed with team and we will work on developing the [error states](https://github.com/cfpb/sbl-frontend/pull/171#issuecomment-1901724082) in a separate ticket.  For now, we want to get the Loading states integrated.

**Review by natalia-fitzgerald — APPROVED** [review]
@meissadia 
Loading is approved. @billhimmelsbach shared a screenshot with me because it wasn't showing up for me.

**Review by billhimmelsbach — APPROVED** [review]
Great! Looking forward to the follow-up component level error PR. 👍

**Comment by meissadia** [comment]
Development of Error state handling moved to [[Design] Error states for failed API calls](https://github.com/cfpb/sbl-frontend/issues/208)
